### DISCUSSION TITLE
Is aux-proof-verify implemented in Solidity yet?

### DISCUSSION THREAD

**Discussion by xhliu** [NilFoundation/evm-mina-verification#47] [Q&A]
Is aux-proof-verify implemented in an Ethereum smart contract yet? Is aux-proof-verify in #28 and thus still being implemented? 

[Readme](https://github.com/NilFoundation/mina-state-proof)  

> 2. An in-EVM application logic aux-proof-verify that has an internal state corresponding to the Mina protocol state, and which can be set to a new state only if one provides an auxiliary proof that verifies.

https://github.com/NilFoundation/evm-placeholder-verification

**Comment by nemothenoone** [comment]
Hey @xhliu! In-EVM verifier is implemented and awaits for an audit. It meanwhile goes through adjustments, fixes and improvements when they're being figured out. The repository you've mentioned is correct.

**Comment by xhliu** [comment]
What’s the gas cost of the verification?

**Reply by nemothenoone** [reply]
A verification cost depends on security parameters which will be picked for production deployment. It can vary from 2,5m gas per state proof to 5+m gas depending on how much data is being hashed into LPC's Merkle tree. We're also constantly adjusting the proof system itself to achieve smaller costs, but this is an ongoing process.
